sycl/include/CL/sycl/detail/pi_args_helper.hpp

```diff
@@ -0,0 +1,540 @@
+//==---------- pi_args_helper.hpp - PI call arguments helper ---------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+/// \file pi_args_helper.hpp
+/// Utilities to enable PI call arguments packing for XPTI.
+///
+/// C++ wrapper for PI does not take real function argument type into account.
```

Comment:
As an alternative approach we could modify `pi.def` to call `_PI_API` macro with APIs take.
Something like https://godbolt.org/z/4W1cYhhKf

Reply:
I like this idea, however in your particular example there's a bug (which was the reason for creating this file in the first place):

consider I want to wait for some events:
```
std::vector<event> es;
Plugin.call<PiApiKind::piEventsWait>(es.size(), es.data());
```

In this case C++ will deduce the template parameter as `size_t`, which is different from `pi_uint32`, used by the actual API. So, we'd store 8 bytes of data when we expect to decode 4.

As a workaround to that problem we could make specializations for packing helper struct through the same macro sorcery.

Reply:
Replaced this file with an extended macro in `pi.def`